hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/ClientClusterServiceImpl.java

```diff
@@ -205,54 +210,38 @@ private ClientConfig getClientConfig() {
         return client.getClientConfig();
     }
 
-    String membersString() {
-        StringBuilder sb = new StringBuilder("\n\nMembers [");
-        final Collection<Member> members = getMemberList();
-        sb.append(members != null ? members.size() : 0);
-        sb.append("] {");
-        if (members != null) {
-            for (Member member : members) {
-                sb.append("\n\t").append(member);
+    void handleMembershipEvent(MembershipEvent event) {
+        synchronized (initialMembershipListenerMutex) {
+            Member member = event.getMember();
+            if (event.getEventType() == MembershipEvent.MEMBER_ADDED) {
+                LinkedHashMap<Address, Member> newMap = new LinkedHashMap<Address, Member>(members.get());
+                newMap.put(member.getAddress(), member);
+                members.set(Collections.unmodifiableMap(newMap));
+            } else {
+                LinkedHashMap<Address, Member> newMap = new LinkedHashMap<Address, Member>(members.get());
+                newMap.remove(member.getAddress());
+                members.set(Collections.unmodifiableMap(newMap));
             }
+
+            fireMembershipEvent(event);
         }
-        sb.append("\n}\n");
-        return sb.toString();
     }
 
-    void fireMembershipEvent(final MembershipEvent event) {
-        ClientExecutionService clientExecutionService = client.getClientExecutionService();
-        clientExecutionService.execute(new Runnable() {
```

Comment:
No, this is not IO thread.
These are event threads. Same as all other listeners callbacks.